src/Memory/InputQueue.ts

```diff
@@ -6,117 +6,149 @@ import * as BB from 'botbuilder'
+                    // Set the input currently being processed (mutex)
+                    await inProcessMessageState.set(inputToProcess)
 
-                // Fire the callback with success
-                CLDebug.Log(`PROCESS-CALLBACK: ${messageProcessing.activityId} ${this.messageQueue.length}`, DebugType.MessageQueue)
-                messageProcessing.callback(false, messageProcessing.activityId)
+                    // Fire the callback with success to start processing
+                    this.log(`CALLBACK `, conversationId, inputToProcess.activityId)
+                    inputToProcess.callback(false, inputToProcess.activityId)
+                }
             }
             else {
-                CLDebug.Log(`PROCESS-ERR: No Message`, DebugType.MessageQueue)
+                // Queue is empty
+                this.log(`FIN QUEUE`, conversationId)
+                delete this.inputQueues[conversationId]
             }
         }
-        else {
-            CLDebug.Log(`PROCESS-NEXT: Empty`, DebugType.MessageQueue)
-        }
     }
 
-    // Done processing message, remove from queue
-    public static async MessageHandled(inProcessMessageState: InProcessMessageState, activityId: string | undefined): Promise<void> {
+    // To be called when an input is done being processeed
+    public static async MessageHandled(inProcessMessageState: InProcessMessageState, conversationId: string, activity?: BB.Activity): Promise<void> {
 
-        if (!activityId) {
-            CLDebug.Log(`HANDLE: Missing activity id`, DebugType.MessageQueue)
-        }
-        let messageProcessing = await inProcessMessageState.remove<QueuedInput>()
+        // Remove mutex
+        let processedInput = await inProcessMessageState.remove<QueuedInput>()
 
-        // Check for consistency
-        if (messageProcessing?.activityId === activityId) {
-            CLDebug.Log(`HANDLE-POP: ${messageProcessing.activityId} ${this.messageQueue.length}`, DebugType.MessageQueue)
+        // When response is pre-empted by an error message, no activity will be passed
+        if (!activity) {
+            CLDebug.Log(`HANDLE-PREEMPTIVE`, DebugType.MessageQueue)
+        }
+        else if (!processedInput) {
+            // Handle called when no input being processed
+            this.log(`NO  MUTEX`, conversationId, activity.id)
+        }
+        else if (processedInput.activityId !== activity.id) {
+            CLDebug.Error("Input Queue: Handle called for different input than one being processed")
+        }
+        else {
+            this.log(`HANDLED  `, conversationId, processedInput.activityId)
         }
 
-        // Process next message in the queue
-        await InputQueue.InputQueueProcessNext(inProcessMessageState)
+        // Process next input in the queue
+        await InputQueue.InputQueueProcessNext(inProcessMessageState, conversationId)
+    }
+
+    private static log(prefix: string, conversationId: string, activityId?: string): void {
```

Comment:
Would use the new typescript syntax to remove repeated variables

Reply:
please provide an example. thx